src/Transport/Receiving/MessagePump.cs

```diff
@@ -322,33 +325,56 @@ await processor.CloseAsync(cancellationToken)
                     }
                 }
-                catch (ServiceBusException onErrorEx) when (onErrorEx.IsTransient || onErrorEx.Reason is ServiceBusFailureReason.MessageLockLost)
+                catch (ServiceBusException onErrorEx) when (onErrorEx.IsTransient ||
+                                                            onErrorEx.Reason is ServiceBusFailureReason.MessageLockLost)
                 {
                     Logger.Debug("Failed to execute recoverability.", onErrorEx);
 
                     await processMessageEventArgs.SafeAbandonMessageAsync(message,
                             transportSettings.TransportTransactionMode,
-                            cancellationToken: messageProcessingCancellationToken)
+                            cancellationToken: processingTokenSource.Token)
                         .ConfigureAwait(false);
                 }
-                catch (Exception onErrorEx) when (onErrorEx.IsCausedBy(messageProcessingCancellationToken))
+                catch (Exception onErrorEx) when (onErrorEx.IsCausedBy(processingTokenSource.Token))
                 {
                     throw;
                 }
                 catch (Exception onErrorEx)
                 {
-                    criticalErrorAction($"Failed to execute recoverability policy for message with native ID: `{message.MessageId}`", onErrorEx, messageProcessingCancellationToken);
+                    // Using messageProcessingCancellationToken to make sure the critical error action can continue
+                    // to execute until the message processing is stopped.
+                    criticalErrorAction(
+                        $"Failed to execute recoverability policy for message with native ID: `{message.MessageId}`",
+                        onErrorEx, messageProcessingCancellationToken);
 
                     await processMessageEventArgs.SafeAbandonMessageAsync(message,
                             transportSettings.TransportTransactionMode,
-                            cancellationToken: messageProcessingCancellationToken)
+                            cancellationToken: processingTokenSource.Token)
                         .ConfigureAwait(false);
                 }
             }
+            finally
+            {
+                processMessageEventArgs.MessageLockLostAsync -= MessageLockLostHandler;
+            }
+
+            async Task MessageLockLostHandler(MessageLockLostEventArgs lockLostArgs)
+            {
+                Logger.Info("Lost the lock while processing message. Cancelling the pipeline execution.", lockLostArgs.Exception);
+                try
+                {
+                    await processingTokenSource.CancelAsync().ConfigureAwait(false);
+                }
+                catch (ObjectDisposedException)
+                {
+                    // ignored
+                    Logger.Info("Lock lost handler executed but cancellation token source was already disposed.", lockLostArgs.Exception);
```

Comment:
Should we bother logging this case?
Not much it can be done here.

Reply:
I think it is worthwhile having some logging in case we ever have to trouble shoot weird races. Should I go down to debug level?

Reply:
Switched to debug for now

Reply:
I also double checked things with our logging guidance and unfortunately we almost always pass the exception to the logger methods and our logging abstraction has no `LevelFormat` overload that enables passing parameters as formats plus the exception. This makes things not useful for structured loggers :(